internal/dataplane/translator/translate_ingress.go

```diff
@@ -81,53 +84,133 @@ func (t *Translator) ingressRulesFromIngressV1() ingressRules {
-		serviceName := fmt.Sprintf(
-			"%s.%s.%s",
-			allDefaultBackends[0].Namespace,
-			defaultBackend.Service.Name,
-			port.CanonicalString(),
-		)
-		service := kongstate.Service{
-			Service: kong.Service{
-				Name: kong.String(serviceName),
-				Host: kong.String(fmt.Sprintf(
-					"%s.%s.%s.svc",
-					defaultBackend.Service.Name,
-					ingress.Namespace,
-					port.CanonicalString(),
-				)),
-				Port:           kong.Int(DefaultHTTPPort),
-				Protocol:       kong.String("http"),
-				ConnectTimeout: kong.Int(DefaultServiceTimeout),
-				ReadTimeout:    kong.Int(DefaultServiceTimeout),
-				WriteTimeout:   kong.Int(DefaultServiceTimeout),
-				Retries:        kong.Int(DefaultRetries),
-				Tags:           util.GenerateTagsForObject(&ingress),
-			},
-			Namespace: ingress.Namespace,
-			Backends: []kongstate.ServiceBackend{{
-				Name:    defaultBackend.Service.Name,
-				PortDef: port,
-			}},
-			Parent: &ingress,
+		defaultBackend := ingress.Spec.DefaultBackend
+		route := translateIngressDefaultBackendRoute(&ingress, util.GenerateTagsForObject(&ingress), features.ExpressionRoutes)
+
+		// If the default backend is defined as an arbitrary resource, then we need handle it differently.
+		if resource := defaultBackend.Resource; resource != nil {
+			return translateIngressDefaultBackendResource(
+				resource,
+				ingress,
+				route,
+				storer,
+				failuresCollector,
+				features,
+			)
 		}
-		r := translateIngressDefaultBackendRoute(&ingress, util.GenerateTagsForObject(&ingress), expressionRoutes)
-		service.Routes = append(service.Routes, *r)
-		return service, true
+
+		// Otherwise, the default backend is defined as a Kubernetes Service.
+		return translateIngressDefaultBackendService(ingress, route)
 	}
 
 	return kongstate.Service{}, false
 }
 
+func translateIngressDefaultBackendResource(
+	resource *corev1.TypedLocalObjectReference,
+	ingress netv1.Ingress,
+	route *kongstate.Route,
+	storer store.Storer,
+	failuresCollector *failures.ResourceFailuresCollector,
+	features FeatureFlags,
+) (kongstate.Service, bool) {
+	if !subtranslator.IsKongServiceFacade(resource) {
+		gk := resource.Kind
+		if resource.APIGroup != nil {
+			gk = *resource.APIGroup + "/" + gk
+		}
+		failuresCollector.PushResourceFailure(fmt.Sprintf("default backend: unknown resource type %s", gk), &ingress)
+		return kongstate.Service{}, false
+	}
+	if !features.KongServiceFacade {
+		failuresCollector.PushResourceFailure(
+			fmt.Sprintf("default backend: KongServiceFacade is not enabled, please set the %q feature gate to 'true' to enable it", featuregates.KongServiceFacade),
+			&ingress,
+		)
+		return kongstate.Service{}, false
+	}
+	facade, err := storer.GetKongServiceFacade(ingress.Namespace, resource.Name)
+	if err != nil {
+		failuresCollector.PushResourceFailure(
+			fmt.Sprintf("default backend: KongServiceFacade %q could not be fetched: %s", resource.Name, err),
```

Comment:
```suggestion
			fmt.Sprintf("default backend: KongServiceFacade %q could not be fetched: %v", resource.Name, err),
```
`%s` and `%v` should be the same but most of the usage in the code base used `%v`. Any reason for changing this?
Or maybe `fmt.Sprintf(".... %s",err.Error())` is another choice?

Reply:
Hmm, I've never really analyzed that, to be honest - `%s` works as expected with non-nil errors like in this case so I didn't really change anything but just used what's correct (the same way as `%v` would be correct 😄).

As for how it's done in the codebase - when I'm searching for `fmt.Sprintf`s with `err` I'd say we have 50/50 for `%s` vs `%v`. In the [thread](https://stackoverflow.com/a/61287626/7958339) on SO it's explained that `%v` might make more sense when you're logging an error that might be nil (then you get prettier `<nil>` instead of ugly `%!s(<nil>)`). If you think we'd benefit from making `%v` used always, I suggest doing this in a separate PR (as there are more usages of `%s` across the codebase).

Reply:
Given that there is no difference for non-nil errors between `%s` and `%v`, it is OK to keep this PR as-is. And there is no guidance about preference of `%s` and `%v`, so we may keep the current status.